File: test/trim_pretty_function.py
"""
This is essentially `grep`.

Search a compiled binary for typenames
and check their surrounding printable characters.
If the compiler successfully optimized away
__PRETTY_FUNCTION__'s prefix and suffix, they
will not appear in the binary.
"""

import sys
import re
import unittest
import os

binary_path = sys.argv.pop()
binary_file = open(binary_path, 'rb')
binary_str = binary_file.read()
typenames = [ 'foo', 'bar::baz', 'bar::quux', 'bar::tpl', 'bar::a_tpl']

def get_prefix_suffix():
  """
  Assume there is one type named cruft_crib
  on which wnaabi::pretty_function::c_str<> is used without
  trimming, and the embedded strings !!!cruft_suffix=1!!!
  !!!cruft_strlen=48!! (where 1 and 48 are the actual lengths
  of the suffix and the whole string respectively)
  to allow capture of the prefix and suffix.
  """
  cruft_strlen = re.search(r'!!!cruft_strlen=(\d+)!!!', binary_str)
  cruft_strlen = int(cruft_strlen.group(1))

  cruft_suffix = re.search(r'!!!cruft_suffix=(\d+)!!!', binary_str)
  cruft_suffix = int(cruft_suffix.group(1))

  cruft_prefix = cruft_strlen - cruft_suffix - len('cruft_crib')

  return re.search(
    '([ -~]{%d})cruft_crib([ -~]{%d})' % (cruft_prefix, cruft_suffix),
    binary_str).groups()

class PrettyFunctionTrimTest(unittest.TestCase):

  def test_prefix_suffix_trimmed(self):
    prefix, suffix = get_prefix_suffix()
    #print "prefix '''%s'''\nsuffix '''%s'''" % (prefix, suffix)
    pattern = ('(.{%d})(' + '|'.join(typenames) + ')(.{%d})'
      ) % (len(prefix), len(suffix))

    for match in re.finditer(pattern, binary_str):
      maybe_prefix, typename, maybe_suffix = match.groups()
      self.assertNotEqual(maybe_prefix, prefix)
      self.assertNotEqual(maybe_suffix, suffix)

if __name__ == "__main__":
  unittest.main()

File: test/trim_pretty_function.cpp
#include <wnaabi/pretty_function.hpp>
#include <iostream>

struct foo;

namespace bar
{
struct baz;

namespace
{
struct quux;

template <typename>
struct a_tpl;
}

template <typename>
struct tpl;
}

using wnaabi::string_literal;
using wnaabi::to_string_literal;
using wnaabi::pretty_function::strlen;
using wnaabi::pretty_function::c_str;

struct cruft_crib;
constexpr auto cruft_suffix =
  string_literal<sizeof("!!!cruft_suffix=")-1>("!!!cruft_suffix=")
  + to_string_literal<char, wnaabi::pretty_function::suffix_length, 10>()
  + string_literal<sizeof("!!!")-1>("!!!");
constexpr auto cruft_strlen =
  string_literal<sizeof("!!!cruft_strlen=")-1>("!!!cruft_strlen=")
  + to_string_literal<char, strlen<cruft_crib>(), 10>()
  + string_literal<sizeof("!!!")-1>("!!!");
constexpr auto cruft_c_str = wnaabi::pretty_function::c_str<cruft_crib>();

constexpr auto name_0 = wnaabi::get_typename<foo>::value;
constexpr auto name_1 = wnaabi::get_typename<bar::baz>::value;
constexpr auto name_2 = wnaabi::get_typename<bar::quux>::value;
constexpr auto name_3 = wnaabi::get_templatename<bar::tpl, int>::value;
constexpr auto name_4 = wnaabi::get_templatename<bar::a_tpl, foo>::value;

int main()
{
  std::cout << cruft_c_str << std::endl;
  std::cout << cruft_strlen << std::endl;
  std::cout << cruft_suffix << std::endl;
  std::cout << name_0 << std::endl;
  std::cout << name_1 << std::endl;
  std::cout << name_2 << std::endl;
  std::cout << name_3 << std::endl;
  std::cout << name_4 << std::endl;
}
